cerbero/build/source.py

```diff
@@ -408,14 +411,93 @@ class Git (GitCache):
     Source handler for git repositories
     '''
 
+    # Maps recipes with common sources share to the same fetch_done condition.
+    _fetch_done_cond_map = {}
+
     def __init__(self):
         GitCache.__init__(self)
         if self.commit is None:
             # Used by recipes in recipes/toolchain/
             self.commit = 'origin/sdk-%s' % self.version
+
+        self._update_fetch_done_cond()
+        self._fetching = False
+
         # For forced commits in the config
         self.commit = self.config.recipe_commit(self.name) or self.commit
 
+    def _get_source_id(self):
+        return (self.repo_dir, tuple(self.remotes.items()))
+
+    def _update_fetch_done_cond(self):
+        source_id = self._get_source_id()
+        if source_id in Git._fetch_done_cond_map:
+            return
+
+        lock = asyncio.Lock()
+        Git._fetch_done_cond_map[source_id] = asyncio.Condition(lock)
+
+    def _get_fetch_done_cond(self):
+        source_id = self._get_source_id()
+        return Git._fetch_done_cond_map[source_id]
+
+    def _get_git_cache_recipes(self):
+        recipes = self.config.cookbook.get_recipes_list()
+        return iter(r for r in recipes if r.stype == SourceType.GIT)
+
+    def _common_source_recipes(self, target_recipe):
+        recipes = self.config.cookbook.get_recipes_list()
+        git_recipes = iter(r for r in recipes if r.stype == SourceType.GIT)
+
+        target_source_id = target_recipe._get_source_id()
+        for recipe in self._get_git_cache_recipes():
+            if target_source_id == recipe._get_source_id():
+                yield recipe
+        yield from ()
+
+    def _already_fetch(self):
```

Comment:
> Now, I think the solution is somehow convoluted and not easy to grasp at first sight. I know what it does is not precisely easy, but I'd like to try push for a simpler implementation. I'm talking particularly about the need of fetch_done, already_fetch and other_fetching functions.

The simpler implementation that I can think of is to address this from the fetch command, instead of the impl.

Reply:
Right. The issue is that the implementation you propose here serves both for fetch (which runs in parallel by default with async) but also for build command which will run fetch if needed. The fetch implementation would be much much simpler because you know beforehands which recipes you are going to fetch. That'd avoid the lock mechanism that you implement here. But of course, on the other side it wouldn't be used for build command.